pkg/controller/cluster/cluster_controller.go

```diff
@@ -35,114 +40,136 @@ import (
-				return reconcile.Result{}, err
-			}
-
-			// Since adding the finalizer updates the object return to avoid later update issues.
-			return reconcile.Result{Requeue: true}, nil
+	if err := r.reconcile(ctx, cluster); err != nil {
+		if requeueErr, ok := errors.Cause(err).(capierrors.HasRequeueAfterError); ok {
+			klog.Infof("Reconciliation for Cluster %q in namespace %q asked to requeue: %v", cluster.Name, cluster.Namespace, err)
+			return reconcile.Result{Requeue: true, RequeueAfter: requeueErr.GetRequeueAfter()}, nil
 		}
-
+		return reconcile.Result{}, err
 	}
 
 	if !cluster.ObjectMeta.DeletionTimestamp.IsZero() {
-		// no-op if finalizer has been removed.
-		if !util.Contains(cluster.ObjectMeta.Finalizers, clusterv1.ClusterFinalizer) {
-			klog.Infof("reconciling cluster object %v causes a no-op as there is no finalizer.", name)
-			return reconcile.Result{}, nil
-		}
-
-		klog.Infof("reconciling cluster object %v triggers delete.", name)
-		if err := r.actuator.Delete(cluster); err != nil {
-			klog.Errorf("Error deleting cluster object %v; %v", name, err)
+		if err := r.isDeleteReady(ctx, cluster); err != nil {
+			if requeueErr, ok := errors.Cause(err).(capierrors.HasRequeueAfterError); ok {
+				klog.Infof("Reconciliation for Cluster %q in namespace %q asked to requeue: %v", cluster.Name, cluster.Namespace, err)
+				return reconcile.Result{Requeue: true, RequeueAfter: requeueErr.GetRequeueAfter()}, nil
+			}
 			return reconcile.Result{}, err
 		}
-		// Remove finalizer on successful deletion.
-		klog.Infof("cluster object %v deletion successful, removing finalizer.", name)
+
 		cluster.ObjectMeta.Finalizers = util.Filter(cluster.ObjectMeta.Finalizers, clusterv1.ClusterFinalizer)
-		if err := r.Client.Update(context.Background(), cluster); err != nil {
-			klog.Errorf("Error removing finalizer from cluster object %v; %v", name, err)
-			return reconcile.Result{}, err
-		}
-		return reconcile.Result{}, nil
 	}
 
-	klog.Infof("reconciling cluster object %v triggers idempotent reconcile.", name)
-	if err := r.actuator.Reconcile(cluster); err != nil {
-		if requeueErr, ok := errors.Cause(err).(capierrors.HasRequeueAfterError); ok {
-			klog.Infof("Actuator returned requeue-after error: %v", requeueErr)
-			return reconcile.Result{Requeue: true, RequeueAfter: requeueErr.GetRequeueAfter()}, nil
+	return reconcile.Result{}, nil
+}
+
+// isDeleteReady returns an error if InfrastructureRef referenced object still exist
+// or if any resources linked to this cluster aren't yet deleted.
+func (r *ReconcileCluster) isDeleteReady(ctx context.Context, cluster *v1alpha2.Cluster) error {
+	// TODO(vincepri): List and delete MachineDeployments, MachineSets, Machines, and
+	// block deletion until they're all deleted.
+
+	if cluster.Spec.InfrastructureRef != nil {
+		_, err := external.Get(r.Client, cluster.Spec.InfrastructureRef, cluster.Namespace)
+		if err != nil && !apierrors.IsNotFound(err) {
```

Comment:
If you wanted, this might look a little cleaner:
```go
switch {
case apierrors.IsNotFound(err):
  // This is what we want - no need to requeue
case err != nil:
  return errors.Wrapf(...)
default: // or case err == nil
  return requeue
}
```